File: server/src/services/attendance.js
const { Op } = require('sequelize');
const { Alumno, Asistencia } = require('../db/models');

async function buscarAlumno(value) {
  let alumno = await Alumno.findOne({ where: { rfid_uid: value } });
  if (!alumno && /^\d+$/.test(value)) {
    alumno = await Alumno.findByPk(parseInt(value));
  }
  return alumno;
}

async function registrarAsistencia(alumno_id) {
  const hoy = new Date();
  hoy.setHours(0, 0, 0, 0);

  const registro = await Asistencia.findOne({
    where: { alumno_id, fecha_entrada: { [Op.gte]: hoy } },
  });

  if (!registro) {
    const nuevo = await Asistencia.create({ alumno_id, fecha_entrada: new Date() });
    return { tipo: 'entrada', registro: nuevo };
  }
  if (!registro.fecha_salida) {
    registro.fecha_salida = new Date();
    await registro.save();
    return { tipo: 'salida', registro };
  }
  return { tipo: 'completo', registro };
}

async function getAsistenciasHoy() {
  const hoy = new Date();
  hoy.setHours(0, 0, 0, 0);

  const registros = await Asistencia.findAll({
    where: { fecha_entrada: { [Op.gte]: hoy } },
    include: [{ model: Alumno, include: ['Grupo'] }],
    order: [['fecha_entrada', 'DESC']],
  });

  return registros.map(r => ({
    alumno: r.Alumno?.nombre ?? '—',
    grupo: r.Alumno?.Grupo ? `${r.Alumno.Grupo.grado}° ${r.Alumno.Grupo.nombre}` : '—',
    fecha_entrada: r.fecha_entrada.toTimeString().slice(0, 8),
    fecha_salida: r.fecha_salida ? r.fecha_salida.toTimeString().slice(0, 8) : '—',
  }));
}

module.exports = { buscarAlumno, registrarAsistencia, getAsistenciasHoy };

File: services/attendance.py
from datetime import datetime, date
from typing import Optional
from db.database import get_session
from db.models import Alumno, Asistencia


def buscar_alumno_por_id(alumno_id: int) -> Optional[Alumno]:
    session = get_session()
    try:
        return session.query(Alumno).filter_by(id=alumno_id).first()
    finally:
        session.close()


def buscar_alumno_por_rfid(rfid_uid: str) -> Optional[Alumno]:
    session = get_session()
    try:
        return session.query(Alumno).filter_by(rfid_uid=rfid_uid).first()
    finally:
        session.close()


def registrar_asistencia(alumno_id: int) -> tuple[str, Asistencia]:
    """
    Detecta automáticamente si es entrada o salida:
    - Si no hay registro hoy → entrada
    - Si hay entrada pero no salida → salida
    - Si ya tiene entrada y salida → retorna aviso sin duplicar

    Retorna (tipo, registro): tipo es 'entrada', 'salida' o 'completo'
    """
    session = get_session()
    try:
        hoy = date.today()
        registro = (
            session.query(Asistencia)
            .filter(
                Asistencia.alumno_id == alumno_id,
                Asistencia.fecha_entrada >= datetime.combine(hoy, datetime.min.time()),
            )
            .first()
        )

        if registro is None:
            nuevo = Asistencia(alumno_id=alumno_id, fecha_entrada=datetime.now())
            session.add(nuevo)
            session.commit()
            session.refresh(nuevo)
            return "entrada", nuevo

        if registro.fecha_salida is None:
            registro.fecha_salida = datetime.now()
            session.commit()
            session.refresh(registro)
            return "salida", registro

        return "completo", registro
    finally:
        session.close()


def get_asistencias_hoy() -> list[dict]:
    session = get_session()
    try:
        hoy = date.today()
        registros = (
            session.query(Asistencia)
            .filter(Asistencia.fecha_entrada >= datetime.combine(hoy, datetime.min.time()))
            .order_by(Asistencia.fecha_entrada.desc())
            .all()
        )

        resultado = []
        for r in registros:
            alumno = session.query(Alumno).filter_by(id=r.alumno_id).first()
            resultado.append({
                "alumno":        alumno.nombre if alumno else "—",
                "grupo":         f"{alumno.grupo.grado}° {alumno.grupo.nombre}" if alumno else "—",
                "fecha_entrada": r.fecha_entrada.strftime("%H:%M:%S"),
                "fecha_salida":  r.fecha_salida.strftime("%H:%M:%S") if r.fecha_salida else "—",
            })
        return resultado
    finally:
        session.close()
